### PR TITLE
feat(dex): derive tick liquidity on demand at T9

### PR DESCRIPTION
**Pull request by decofe**
Stacked on #6614. Implements the remaining TIP-1024 tick-liquidity changes at T9.

## What
- Stops maintaining the stored per-tick `totalLiquidity` aggregate on T9 for placements, partial/full fills, and cancellations.
- Computes `getTickLevel.totalLiquidity` on demand by walking the active order list with checked arithmetic.
- Leaves pre-T9 behavior unchanged for historical replay and legacy quote paths; pre-existing aggregate slots become harmless stale state at T9.
- Adds a fork-transition test covering the maintained pre-T9 value, equal T9 boundary value, and correct T9 derivation after the raw slot becomes stale.

Spec update: #6940

## Validation
- `cargo +nightly fmt --all -- --check`
- `cargo metadata --no-deps --format-version 1`
- `git diff --check`
- `cargo test -p tempo-precompiles stablecoin_dex --lib` blocked before compilation: Cargo received HTTP 401 fetching pinned `commonware-broadcast` from `commonwarexyz/monorepo`.

Prompted by: @0xrusowsky

### CODE REVIEW

File: crates/precompiles/src/stablecoin_dex/mod.rs [hunk 1]
```diff
@@ -713,6 +732,18 @@ impl StablecoinDEX {
...
+            .spec()
+            .is_t9()
+            .then(|| self.compute_tick_liquidity(level.head))
+            .transpose()?;
+        if let Some(liquidity) = t9_liquidity {
+            // Preserve the placement-time overflow guard before making any linked-list writes.
+            liquidity
+                .checked_add(order.remaining())
+                .ok_or(TempoPrecompileError::under_overflow())?;
+        }
```
Comment (0xrusowsky) [inline crates/precompiles/src/stablecoin_dex/mod.rs:732]: rewrite this is a comprehensive manner, clearly branching on an "if is_t9()"

File: crates/precompiles/src/stablecoin_dex/mod.rs [hunk 2]
```diff
@@ -1030,16 +1062,16 @@ impl StablecoinDEX {
         let amount_out = taker_output(fill_amount, order.tick(), order.is_bid())
             .ok_or(TempoPrecompileError::under_overflow())?;
 
-        // Update price level total liquidity
```
Comment (0xrusowsky) [inline crates/precompiles/src/stablecoin_dex/mod.rs:1033]: preserve comments

File: crates/precompiles/src/stablecoin_dex/mod.rs [hunk 3]
```diff
@@ -1352,12 +1385,12 @@ impl StablecoinDEX {
             level.tail = order.prev();
         }
 
-        // Update level liquidity
```
Comment (0xrusowsky) [inline crates/precompiles/src/stablecoin_dex/mod.rs:1355]: preserve comments

**Comment by 0xrusowsky** [comment]
cyclops audit

**Comment by decofe** [comment]
cc @0xrusowsky

Cyclops audit event published. [View workflow run](https://github.com/tempoxyz/tempo/actions/runs/29973487617)

**Config:** config: `default`, iterations: `default`, hours: `default`

**Review by legion2002 — APPROVED** [review]

**Comment by 0xrusowsky** [comment]
cyclops audit fast note="assess whether 8bdd13cc02f22fe634851d776614248054490f6d fixes https://github.com/tempoxyz/tempo/pull/6939#discussion_r3635413709"

**Comment by decofe** [comment]
cc @0xrusowsky

Cyclops audit event published. [View workflow run](https://github.com/tempoxyz/tempo/actions/runs/30107637774)

**Config:** config: `default`, iterations: `1`, hours: `default`, note: `assess whether 8bdd13cc02f22fe634851d776614248054490f6d fixes https://github.com/tempoxyz/tempo/pull/6939#discussion_r3635413709`

File: crates/precompiles/src/stablecoin_dex/mod.rs [hunk 4]
```diff
@@ -713,6 +729,12 @@ impl StablecoinDEX {
         let mut level = self.books[order.book_key()]
             .tick_level_handler(order.tick(), order.is_bid())
             .read()?;
+        if !self.storage.spec().is_t9() {
```
[the remark this answers is not in the document]
Comment (0xrusowsky) [reply crates/precompiles/src/stablecoin_dex/mod.rs]: addressed in https://github.com/tempoxyz/tempo/pull/6939/commits/8bdd13cc02f22fe634851d776614248054490f6d

File: crates/precompiles/src/stablecoin_dex/mod.rs [hunk 5]
```diff
@@ -713,6 +729,12 @@ impl StablecoinDEX {
...
+            level.total_liquidity = level
+                .total_liquidity
+                .checked_add(order.remaining())
+                .ok_or(TempoPrecompileError::under_overflow())?;
```
Comment (0xKitsune) [inline crates/precompiles/src/stablecoin_dex/mod.rs]: For clarity, why did we move this up?

File: crates/precompiles/src/stablecoin_dex/mod.rs (hunk 4 above)
Comment (0xKitsune) [inline crates/precompiles/src/stablecoin_dex/mod.rs]: Could we keep the pre-T9 order here?

Before this PR we linked the previous tail before checking the liquidity overflow. This now checks first. The state reverts either way, but the storage work is metered, so historical replay can use different gas.

Can we preserve the old ordering before T9 and only skip the aggregate update at T9?

File: crates/precompiles/src/stablecoin_dex/mod.rs [hunk 6]
```diff
@@ -747,12 +769,6 @@ impl StablecoinDEX {
...
 
-        let new_liquidity = level
-            .total_liquidity
-            .checked_add(order.remaining())
-            .ok_or(TempoPrecompileError::under_overflow())?;
-        level.total_liquidity = new_liquidity;
-
         self.books[order.book_key()]
             .tick_level_handler_mut(order.tick(), order.is_bid())
             .write(level)?;
```
Comment (0xKitsune) [inline crates/precompiles/src/stablecoin_dex/mod.rs:775]: Could we avoid writing the full TickLevel at T9?

TickLevel::write writes both storage slots, including totalLiquidity. The full read above also loads it. The same full write is used when advancing after a full fill and when cancelling. So T9 still pays for the storage write this change is meant to remove.

Can the T9 path read and write only head and tail?

File: crates/precompiles/src/stablecoin_dex/mod.rs (hunk 5 above)
Comment (0xrusowsky) [reply crates/precompiles/src/stablecoin_dex/mod.rs]: addressed in https://github.com/tempoxyz/tempo/commit/8438b7244090acbefafac871197e91085320aa44

File: crates/precompiles/src/stablecoin_dex/mod.rs (hunk 4 above)
Comment (0xrusowsky) [reply crates/precompiles/src/stablecoin_dex/mod.rs]: addressed in https://github.com/tempoxyz/tempo/commit/8438b7244090acbefafac871197e91085320aa44

File: crates/precompiles/src/stablecoin_dex/mod.rs (hunk 6 above)
Comment (0xrusowsky) [reply crates/precompiles/src/stablecoin_dex/mod.rs:775]: addressed in https://github.com/tempoxyz/tempo/commit/8438b7244090acbefafac871197e91085320aa44

File: crates/precompiles/src/stablecoin_dex/mod.rs [hunk 7]
```diff
@@ -500,11 +500,27 @@ impl StablecoinDEX {
...
+            self.books[book_key].asks[tick].read()?
+        };
+
+        if self.storage.spec().is_t9() {
+            // Sum the remaining amount of every order reachable from the tick's head.
+            let mut order_id = level.links.head;
+            level.total_liquidity = 0;
+            while order_id != 0 {
+                let order = self.orders[order_id].read_in_book(book_key)?;
+                level.total_liquidity = level
```
Comment (mattsse) [inline crates/precompiles/src/stablecoin_dex/mod.rs:515]: Could we add a regression for this overflow branch itself? At T9, place two valid same-tick orders whose combined `remaining` exceeds `u128::MAX` (both placements should succeed), then call `getTickLevel` and assert the under/overflow revert while also confirming the linked-list head and tail are unchanged. The existing fork-transition test only exercises representable sums, so wrapping/saturating arithmetic or different error mapping could regress without being detected.

**Review by mattsse — COMMENTED** [review]
The implementation matches the T9 on-demand-liquidity requirements. I found one non-blocking regression-test gap in the new consensus-facing overflow path.

File: crates/precompiles/src/stablecoin_dex/mod.rs (hunk 7 above)
Comment (0xrusowsky) [reply crates/precompiles/src/stablecoin_dex/mod.rs:515]: addressed in https://github.com/tempoxyz/tempo/commit/80e2e8ccd0
